### PR TITLE
Implement ESIM

### PR DESCRIPTION
**Pull request by jellying**
Implement ESIM

### CODE REVIEW

**Comment by jellying** [comment]
#667 Implement ESIM

File: matchzoo/models/esim.py [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+        #   R = `input_right` sequence length
+        #   H = LSTM hidden size
+
+        # Left input and right input
+        # shape = [B, L]
+        # shape = [B, R]
+        input_left, input_right = self._make_inputs()
+        hidden_size = self._params['lstm_units']
+
+        def tile_tensor(x):
```
Comment (bwanglzu) [inline matchzoo/models/esim.py]: put these functions into a common namespace for reuse,

File: matchzoo/models/esim.py [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+        ))
+        # Encode the two text
+        # shape = [B, L, 2*H]
+        encoded_left = encoder_layer(embed_left)
+        # shape = [B, R, 2*H]
+        encoded_right = encoder_layer(embed_right)
+        # Make interaction and get the interactive representation
+        # shape = [B, L, 2*H]
+        # shape = [B, R, 2*H]
+        attended_left, attended_right = self._attention(encoded_left, left_mask, encoded_right, right_mask)
```
Comment (bwanglzu) [inline matchzoo/models/esim.py]: `encoded_left` `encoded_right` 

`left_mask` `right_mask`?  -> `mask_left` `mask_right`

at least keep your variable names consistent internally

File: matchzoo/models/esim.py [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+        # shape = [B, L, 8*H]
+        enhanced_left = keras.layers.Concatenate()([encoded_left, attended_left,
+                                                    keras.layers.Subtract()([encoded_left, attended_left]),
+                                                    keras.layers.Multiply()([encoded_left, attended_left])])
```
Comment (bwanglzu) [inline matchzoo/models/esim.py]: I guess you do not need to  call `keras.layers.Substract` or `keras.layers,Multiply` or even `keras.layers.ReLU()` explicitly, use `substract`, `multiply` and `relu`  (string format) directly, which is more readable.

File: matchzoo/models/esim.py [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+        # shape = [B, R, 2*H]
+        fusion_right = fusion_layer(projected_right)
+        # Do global avg and max pooling and get a fixed-length representation vector
+        avg_left_rep = keras.layers.Lambda(lambda x: K.sum(x * expand_tensor(left_mask), axis=1) /
+                                                     expand_tensor(K.sum(left_mask, axis=1)))(fusion_left)
+        avg_right_rep = keras.layers.Lambda(lambda x: K.sum(x * expand_tensor(right_mask), axis=1) /
+                                                      expand_tensor(K.sum(right_mask, axis=1)))(fusion_right)
+        max_left_rep = keras.layers.Lambda(lambda x: K.max(x + (1 - tile_tensor(left_mask)) * -1e7, axis=1))(
+            fusion_left)
+        max_right_rep = keras.layers.Lambda(lambda x: K.max(x + (1 - tile_tensor(right_mask)) * -1e7, axis=1))(
```
Comment (bwanglzu) [inline matchzoo/models/esim.py]: can we set `-1e7` as an epsilon paramterer?

File: matchzoo/models/esim.py [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+            fusion_right)
+        # shape = [B, 8*H]
+        cls_input = keras.layers.Concatenate()([avg_left_rep, avg_right_rep, max_left_rep, max_right_rep])
```
Comment (bwanglzu) [inline matchzoo/models/esim.py]: what is `cls`?

File: matchzoo/models/esim.py [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+        # Interaction
+        embed_cross = keras.layers.Dot(axes=2)([encoded_left, encoded_right])
+        cross_shape = [self._params['input_shapes'][0][0], self._params['input_shapes'][1][0]]
+        embed_cross = keras.layers.Reshape(cross_shape)(embed_cross)
+        left2right_attn = self._masked_softmax(embed_cross, left_mask,
+                                               self._params['input_shapes'][0][0],
+                                               self._params['input_shapes'][1][0])
+        # left2right_attn = keras.layers.Lambda(lambda x: K.tf.Print(x, [x]))(left2right_attn)
+
+        right2left_attn = self._masked_softmax(keras.layers.Permute([2, 1])(embed_cross), right_mask,
```
Comment (bwanglzu) [inline matchzoo/models/esim.py]: I don't like the way you naming your variables, specifically, digits together with letters.

File: matchzoo/models/esim.py [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+                                            encoded_left,
+                                            right_mask)
+        return attended_left, attended_right
+
+    def _masked_softmax(self, input, mask, att_len, base_len):
+        inf = 1e8
+
+        def tile_mask(x):
+            res = inf * K.tile(K.expand_dims(1 - x, axis=1), [1, base_len, 1])
+            return res
```
Comment (bwanglzu) [inline matchzoo/models/esim.py]: same

File: matchzoo/models/esim.py [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+
+    def _weighted_sum(self, weights, tensor, mask):
+        weighted_sum = keras.layers.Dot(axes=[2, 1])([weights, tensor])
+        expanded_mask = keras.layers.Lambda(expand_tensor)(mask)
+        return keras.layers.Multiply()([expanded_mask, weighted_sum])
+
+
+def get_mask(x):
+    boolean_mask = K.not_equal(x, 0)
+    return K.cast(boolean_mask, K.dtype(x))
```
Comment (bwanglzu) [inline matchzoo/models/esim.py]: let others reuse this piece of code

File: matchzoo/models/esim.py [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+
+
+def expand_tensor(x):
```
Comment (bwanglzu) [inline matchzoo/models/esim.py]: let others reuse this piece of code

File: matchzoo/models/esim.py [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+    return K.expand_dims(x)
+
+
+def sum_tensor(x):
```
Comment (bwanglzu) [inline matchzoo/models/esim.py]: let others reuse this piece of code

File: matchzoo/models/esim.py (hunk 8 above)
Comment (jellying) [reply matchzoo/models/esim.py]: These piece of code are very brief. Should I package the each function as a layer like MatchingLayer?

Comment (uduse) [reply matchzoo/models/esim.py]: I think keeping this local is perfectly fine. We could refactor it later when there are more uses.

File: matchzoo/models/esim.py (hunk 4 above)
Comment (jellying) [reply matchzoo/models/esim.py]: This layer is  equal to masked 1d-MaxPooling，-1e7  option is to set the padding position's value to -inf.

File: matchzoo/contrib/models/esim.py [hunk 11]
```diff
@@ -0,0 +1,234 @@
...
+        projected_right = projection(enhanced_right)
+        projected_left = keras.layers.Dropout(
+            rate=self._params['dropout_rate'])(projected_left)
+        projected_right = keras.layers.Dropout(
+            rate=self._params['dropout_rate'])(projected_right)
+        # Fusion the enhanced features by another BiLSTM
+        fusion_layer = keras.layers.Bidirectional(keras.layers.LSTM(
+            hidden_size,
+            return_sequences=True,
+            dropout=self._params['dropout_rate']
```
Comment (pl8787) [inline matchzoo/contrib/models/esim.py]: No dropout here.

File: matchzoo/contrib/models/esim.py [hunk 12]
```diff
@@ -0,0 +1,234 @@
...
+        enhanced_right = keras.layers.concatenate([encoded_right, attended_right,
+                                                   keras.layers.subtract([encoded_right, attended_right]),
+                                                   keras.layers.multiply([encoded_right, attended_right])])
+        # Project to the H size
+        projection = keras.layers.Dense(hidden_size, activation=keras.layers.ReLU())
+        # shape = [B, L, H]
+        projected_left = projection(enhanced_left)
+        # shape = [B, R, H]
+        projected_right = projection(enhanced_right)
+        projected_left = keras.layers.Dropout(
```
Comment (pl8787) [inline matchzoo/contrib/models/esim.py]: RNNDropout

File: matchzoo/contrib/models/esim.py [hunk 13]
```diff
@@ -0,0 +1,234 @@
...
+    #     softmax_res = keras.layers.Softmax() \
+    #         (keras.layers.multiply([reshaped_input, reshaped_mask]))
+    #     softmax_res = keras.layers.multiply([softmax_res, reshaped_mask])
+    #     softmax_res = keras.layers.Lambda(lambda x: x / K.sum(x, axis=-1, keepdims=True) + epsilon)(softmax_res)
+    #     output = keras.layers.Lambda(lambda x: K.reshape(x, [-1, att_len, base_len]))(softmax_res)
+    #     output = keras.layers.Lambda(lambda x: K.tf.Print(x, [x]))(output)
+    #     return output
+
+    def _masked_softmax(self, input, mask, att_len, base_len):
+        inf = 1e8
```
Comment (pl8787) [inline matchzoo/contrib/models/esim.py]: 1e8 and 1e7 define as the same parameter on the top of the model file.
such as `INF = 1e7`
